state/entrysync.go

```diff
@@ -0,0 +1,313 @@
+}
+
+// check routine, for description see EntrySync comment
+func (es *EntrySync) check() {
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+
+		if len(es.eblocks) == 0 { // outside of mutex but should be fine
+			time.Sleep(time.Second)
+			continue
+		}
+
+		es.ebMtx.Lock()
+		eb := es.eblocks[0]
+		es.s.SetEntryBlockDBHeightProcessing(eb.height)
+
+		stillmissing := make([]interfaces.IHash, 0, len(eb.missing))
+		for _, entryhash := range eb.missing {
+			if !es.has(entryhash) {
+				stillmissing = append(stillmissing, entryhash)
+			}
+		}
+
+		if len(stillmissing) > 0 {
+			eb.missing = stillmissing
+		} else { // eblock is complete
+			es.eblocks = es.eblocks[1:]
+			es.s.SetEntryBlockDBHeightComplete(eb.height)
+		}
+		es.ebMtx.Unlock()
+
+		if len(stillmissing) > 0 { // waiting on requests
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+// has checks if the hash exists in the database
+func (es *EntrySync) has(hash interfaces.IHash) bool {
+	if has, err := es.s.DB.DoesKeyExist(databaseOverlay.ENTRY, hash.Bytes()); err != nil {
+		panic(err)
+	} else {
+		return has
+	}
+}
+
+// ask routine, for description see EntrySync comment
+func (es *EntrySync) ask() {
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+
+		es.askMtx.Lock()
+		// check the first N entries in the list
+		for i := 0; i < len(es.asks) && i < EntrySyncMax; i++ {
+			ask := es.asks[i]
+
+			// remove asks that are fulfilled
+			if es.has(ask.hash) {
+				es.asks = append(es.asks[:i], es.asks[i+1:]...)
+				i-- // we removed one, index stays the same
+				delete(es.askMap, ask.hash.Fixed())
+				continue
+			}
+
+			// re-ask
+			if time.Since(ask.time) > EntrySyncRetry {
+				request := messages.NewMissingData(primitives.NewTimestampNow(), ask.hash)
+				request.SendOut(es.s, request)
+				ask.time = time.Now()
+				ask.tries++
+
+				if ask.tries%EntrySyncWarning == 0 {
+					es.s.Logger.WithField("tries", ask.tries).WithField("hash", ask.hash.String()).Warnf("Unable to retrieve entry from network")
+				}
+			}
+		}
+		es.askMtx.Unlock()
+
+		time.Sleep(time.Second)
+	}
+}
+
+// returns the maximum possible height we have eblocks for
+func (es *EntrySync) syncMax() uint32 {
+	end := es.s.GetHighestSavedBlk()
+	if es.s.DBStates.ProcessHeight < end {
+		end = es.s.DBStates.ProcessHeight
+	}
+	return end
+}
+
+// SyncHeight starts the Sync, Ask, and Check routines.
+func (es *EntrySync) SyncHeight() {
+	go es.ask()
+	go es.check()
+
+	es.position = es.s.EntryDBHeightComplete
+	// genesis block edge case since EntryDBHeightComplete can't be -1
+	if es.position > 0 {
+		es.position++
+	}
+
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+		// block not available yet
+		if es.position > es.syncMax() {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		db := es.s.GetDirectoryBlockByHeight(es.position)
+		if db == nil { // block not saved yet
+			time.Sleep(time.Second)
+			continue
+		}
+
+		// legacy logic
```

Comment:
The intent was that someone who wanted to know why that code was there could look at the old version to see what it did, but you're right that doesn't really help anyone. I included a paraphrase of the original comment.

Sidenote: what do you think about that functionality? It's supposed to interrupt the for{} loop to prevent golang from running it uninterrupted, so that it can focus on loading the database. The original code had a sleep of 1 millisecond, which I'm not convinced does much:
https://github.com/FactomProject/factomd/blob/21a1980ab66447018afed2f7d42ac7799239281b/state/entrySyncing.go#L269-L272

The canonical way to interrupt tight loops would be using `runtime.Gosched()` but it's tricky to benchmark this sort of stuff.

Reply:
Are you concerned that `runtime.Gosched()` would have more performance impact than the current `time.Sleep(time.Millisecond * 10)`? 
Given the explicit purpose of the sleep here is to give a chance to other goroutines to run, and that's exactly what `runtime.Gosched()` is supposed to do, I'd rather go that way than using sleep as a workaround, no?

Reply:
I think it's a choice between runtime.Gosched or maybe just delaying execution entirely until the database is finished loading. I'm not sure there's a good reason to have second pass sync running while the node is booting up.